### ISSUE TITLE
unsupported operation SquaredDifference & Pow

### ISSUE THREAD

**Issue by pribadihcr** [tensorflow/tensorflow#19021]
Hi,

There are some unsupported operation

2018-05-02 16:44:35.396116: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396179: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.396269: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396311: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.396454: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396494: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.396635: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396675: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.396792: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396818: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.396909: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.396936: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.397025: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.397051: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow
2018-05-02 16:44:35.397145: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: SquaredDifference
2018-05-02 16:44:35.397172: I tensorflow/contrib/lite/toco/import_tensorflow.cc:1270] Converting unsupported operation: Pow

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gbildson** [comment]
Endorse.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gbildson** [comment]
I see pow in r1.10 and I have a way to get rid of the SquaredDifference in moments so yeah!

**Comment by wincle** [comment]
@gbildson Thx, you inspired me!
In my situation , I changed the tensorflow code and fix it!
change the code in tensorflow/python/ops/nn_impl.py
`math_ops.squared_difference(y, array_ops.stop_gradient(mean)),`
to
'math_ops.square(math_ops.subtract(y, array_ops.stop_gradient(mean))),'
which is in mements def.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
Is SquaredDifference being used in your training graph or eval graph. Right now you should only try to convert an eval graph which should probably not have the squared difference operation (commonly used in the loss). So create a separate eval graph and freeze the training checkpoint into that. 

If SquaredDifference is needed in your eval graph, ignore my above comment. And I have tracked your ask of squared difference in the aggregation issue above.

**Comment by gbildson** [comment]
I'm using InstanceNormalization in a trained Image Transformation Network.  This uses tf.moments to do it's thing so I do need it.  My realization and workaround was that I can replace the squared_difference in tf.moments with a tf.subtraction and tf.multiply to do the same.  This appears to work although may not be as performant. 
Thanks-greg
    On Saturday, August 18, 2018, 8:39:36 PM EDT, Suharsh Sivakumar <notifications@github.com> wrote:  
 
 
Is SquaredDifference being used in your training graph or eval graph. Right now you should only try to convert an eval graph which should probably not have the squared difference operation (commonly used in the loss). So create a separate eval graph and freeze the training checkpoint into that.

If SquaredDifference is needed in your eval graph, ignore my above comment. And I have tracked your ask of squared difference in the aggregation issue above.

**Comment by vholod** [comment]
dear gbildson,
Can you please explain how did you bypassed the usage of tf.moments (in order to calculate the var and then var^0.5 for the instance normalization )?

Thanks

**Comment by gbildson** [comment]
Let me try that more cleanly ...
I went into the tf.moments source and changed it to use supported methods:
`    diff = math_ops.subtract(y, array_ops.stop_gradient(mean))
    variance = math_ops.reduce_mean(
        #math_ops.squared_difference(y, array_ops.stop_gradient(mean)),
        math_ops.multiply(diff,diff),
        axes,
        keepdims=True,
        name="variance")`

and in recent tflite versions, pow is supported so this works in instance normalization:
`            normalized = (inputs - mu) / (sigma_sq + EPSILON) ** .5
`

**Comment by gbildson** [comment]
Now I'm just wondering about the performance and memory overhead of that manual squared_difference.

**Comment by vholod** [comment]
Thanks @gbildson ,
I can try to benchmark the model with those changes, and I will try to update. My last solution did use a lot of memory to perform Square operations:

```
2018-10-24 15:35:10.255607: I tensorflow/core/util/stat_summarizer.cc:85] ============================== Summary by node type ==============================
2018-10-24 15:35:10.255611: I tensorflow/core/util/stat_summarizer.cc:85]                    [Node type]          [count]         [avg ms]         [avg %]          [cdf %]       [mem KB]      [times called]
2018-10-24 15:35:10.255616: I tensorflow/core/util/stat_summarizer.cc:85]                         Conv2D               14         5227.149         76.688%          76.688%     1152921.625            14
2018-10-24 15:35:10.255620: I tensorflow/core/util/stat_summarizer.cc:85]            Conv2DBackpropInput                2          614.725          9.019%          85.707%     1293926.375             2
2018-10-24 15:35:10.255625: I tensorflow/core/util/stat_summarizer.cc:85]                           Mean               64          283.185          4.155%          89.861%       7431.704             64
2018-10-24 15:35:10.255629: I tensorflow/core/util/stat_summarizer.cc:85]                            Mul               33          150.679          2.211%          92.072%          0.000             33
2018-10-24 15:35:10.255633: I tensorflow/core/util/stat_summarizer.cc:85]                            Add               32          142.681          2.093%          94.165%          0.000             32
2018-10-24 15:35:10.255642: I tensorflow/core/util/stat_summarizer.cc:85]                            Sub               16          133.475          1.958%          96.123%          0.000             16
2018-10-24 15:35:10.255649: I tensorflow/core/util/stat_summarizer.cc:85]                         Square               16          126.811          1.860%          97.984%     1551052.750            16
2018-10-24 15:35:10.255657: I tensorflow/core/util/stat_summarizer.cc:85]                           Relu               10           82.338          1.208%          99.192%          0.000             10
2018-10-24 15:35:10.255664: I tensorflow/core/util/stat_summarizer.cc:85]                           AddN                5           52.525          0.771%          99.962%          0.000              5
2018-10-24 15:35:10.255675: I tensorflow/core/util/stat_summarizer.cc:85]                        Sigmoid                1            2.446          0.036%          99.998%          0.000              1
2018-10-24 15:35:10.255682: I tensorflow/core/util/stat_summarizer.cc:85]                          Const               54            0.069          0.001%          99.999%          0.000             54
2018-10-24 15:35:10.255690: I tensorflow/core/util/stat_summarizer.cc:85]                          Rsqrt               16            0.043          0.001%         100.000%          0.000             16
2018-10-24 15:35:10.255697: I tensorflow/core/util/stat_summarizer.cc:85]                        _Retval                1            0.005          0.000%         100.000%          0.000              1
2018-10-24 15:35:10.255705: I tensorflow/core/util/stat_summarizer.cc:85]                           NoOp                1            0.004          0.000%         100.000%          0.000              1
2018-10-24 15:35:10.255714: I tensorflow/core/util/stat_summarizer.cc:85]                           _Arg                1            0.001          0.000%         100.000%          0.000              1
```
